docs/rfcs/GPUXDialect.md

```diff
@@ -0,0 +1,176 @@
+
+## Proposal
+
+We propose to have the GPUX dialect with ops listed below:
+
+### Operation:
+
+#### gpux.create_stream (gpux::CreateStreamOp)
+
+Create the GPU Stream.
+
+Syntax:
+
+```
+operation ::= "gpux.create_stream"() : () -> !gpux.StreamType
+```
+
+The gpux.create_stream() operation will be added by a custom pass only if user does not want to use their own stream.
+This operation will be converted to runtime call for a temporary stream creation in the LLVM generated code. The runtime routine will create a SYCL/L0 queue, device and context based on the runtime.
+
+
+#### gpux.destroy_stream (gpux::DestroyStreamOp)
+
+Destroy the GPU Stream.
+
+Syntax:
+
+```
+operation ::= "gpux.destroy_stream"() : () -> !gpux.StreamType
+```
+
+The gpux.destroy_stream() operation will destroy the above created stream. If the user wants to use their stream, we expect the user to manage the lifetime of the stream.
+
+
+#### gpux.alloc (gpux::AllocOp)
+
+GPU memory allocation operation.
+
+Syntax:
+
+```
+operation ::= gpux.alloc custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
+              ($gpux_stream, $dynamicSizes) ([ $symbolOperands^ ])? attr-dict : type($memref)
+```
+
+This op is an extension of upstream gpu.alloc op with one added argument for stream.
+
+
+#### gpux.dealloc (gpux::DeallocOp)
+
+GPU memory deallocation operation.
+
+Syntax:
+
+```
+operation ::= gpux.dealloc custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
+              ($gpux_stream, $memref) attr-dict : type($memref)
+```
+
+This op is an extension of upstream gpu.dealloc op with one added argument for stream.
+
+
+#### gpux.launch_func (gpux::LaunchFuncOp)
+
+Launches a function as a GPU kernel.
+
+Syntax:
+
+```
+operation ::= gpux.launch_func custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
+              $kernel
+              blocks in  ($gridSizeX, $gridSizeY, $gridSizeZ)
+              threads in ($blockSizeX, $blockSizeY, $blockSizeZ)
+              (dynamic_shared_memory_size $dynamicSharedMemorySize^)?
+              custom<LaunchFuncOperands>($gpux_stream, $operands, type($operands)) attr-dict
```

Comment:
Why $gpux_stream is part of device function parameter? It should be part of GPUX.launch_func(), so becomes the host code interacting with runtime, not the device kernel.

Reply:
What is the reason we don't support gpu.launch also?

Reply:
yes, thats a mistake. it has to be next to kernel. will change it

We dont support gpu.launch is because we assume the user to run the upstream GpuKernelOutliningPass which would translate the gpu.launch to gpu.launch_func and outline the device kernel. Most of the lower passes expect the kernel to be outlined